File: 14-TypeScript/Grey Multi-Tenant/packages/grey-matlab/+grey/+domain/AuthClient.m
classdef AuthClient < handle
    %AUTHCLIENT Authentication client for Grey SDK
    %   Handles login, logout, and token refresh operations.
    
    properties (SetAccess = private)
        httpClient  % grey.http.HttpClient
    end
    
    methods
        function obj = AuthClient(httpClient)
            %AUTHCLIENT Construct auth client
            %   client = AuthClient(httpClient)
            
            obj.httpClient = httpClient;
        end
        
        function result = login(obj, username, password)
            %LOGIN Authenticate with username and password
            %   result = client.login(username, password)
            
            % Validate inputs
            if isempty(strtrim(username))
                result = grey.error.Result.failure(...
                    grey.error.GreyError.validation('Username is required'));
                return;
            end
            
            if isempty(strtrim(password))
                result = grey.error.Result.failure(...
                    grey.error.GreyError.validation('Password is required'));
                return;
            end
            
            body = struct('username', username, 'password', password);
            result = obj.httpClient.post('/auth/login', body);
            
            % Set auth token if successful
            if result.isOk()
                data = result.unwrap();
                if isstruct(data) && isfield(data, 'access_token')
                    obj.httpClient.setAuthToken(data.access_token);
                end
            end
        end
        
        function result = logout(obj)
            %LOGOUT Log out the current user
            %   result = client.logout()
            
            result = obj.httpClient.post('/auth/logout');
            
            % Clear token regardless of result
            obj.httpClient.clearAuthToken();
        end
        
        function result = refresh(obj, refreshToken)
            %REFRESH Refresh the access token
            %   result = client.refresh(refreshToken)
            
            % Validate inputs
            if isempty(strtrim(refreshToken))
                result = grey.error.Result.failure(...
                    grey.error.GreyError.validation('Refresh token is required'));
                return;
            end
            
            body = struct('refresh_token', refreshToken);
            result = obj.httpClient.post('/auth/refresh', body);
            
            % Set new auth token if successful
            if result.isOk()
                data = result.unwrap();
                if isstruct(data) && isfield(data, 'access_token')
                    obj.httpClient.setAuthToken(data.access_token);
                end
            end
        end
    end
end

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-julia/src/AuthClient.jl
"""
Authentication client for the Grey SDK.
"""
module AuthClient

export login, logout, refresh

using ..HttpClient
using ..GreyError
using ..Results

"""
    login(client::HttpClient.Client, username::String, password::String) -> Result

Authenticate with username and password.
Returns tokens on success.
"""
function login(client::HttpClient.Client, username::String, password::String)
    # Validate inputs
    if isempty(strip(username))
        return Results.err(GreyError.validation("Username is required"))
    end
    
    if isempty(strip(password))
        return Results.err(GreyError.validation("Password is required"))
    end
    
    result = HttpClient.post(client, "/auth/login", Dict(
        "username" => username,
        "password" => password
    ))
    
    if Results.is_ok(result) && result.data !== nothing
        token = get(result.data, "access_token", nothing)
        if token !== nothing
            HttpClient.set_auth_token!(client, token)
        end
    end
    
    result
end

"""
    logout(client::HttpClient.Client) -> Result

Log out the current user.
"""
function logout(client::HttpClient.Client)
    result = HttpClient.post(client, "/auth/logout")
    
    # Clear token regardless of result
    HttpClient.clear_auth_token!(client)
    
    result
end

"""
    refresh(client::HttpClient.Client, refresh_token::String) -> Result

Refresh the access token using a refresh token.
"""
function refresh(client::HttpClient.Client, refresh_token::String)
    # Validate inputs
    if isempty(strip(refresh_token))
        return Results.err(GreyError.validation("Refresh token is required"))
    end
    
    result = HttpClient.post(client, "/auth/refresh", Dict(
        "refresh_token" => refresh_token
    ))
    
    if Results.is_ok(result) && result.data !== nothing
        token = get(result.data, "access_token", nothing)
        if token !== nothing
            HttpClient.set_auth_token!(client, token)
        end
    end
    
    result
end

end # module
